### PR TITLE
Add new options

### PR DESCRIPTION
**Pull request by makkaronis4e**
Add comments, login-form, login/logout functionality.

### CODE REVIEW

**Comment by makkaronis4e** [comment]
@webknjaz done

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
+    "python.pythonPath": "${workspaceFolder}/myvenv/bin/python",
+    "python.formatting.provider": "yapf"
```
Comment (webknjaz) [inline .vscode/settings.json]: this one is horrible

File: blog/views.py [hunk 2]
```diff
@@ -56,3 +55,70 @@ class EditPost(UpdateView, Protected):
...
+
+class PostDraftList(ListView, Protected):
+    """Return draft list."""
+
+    queryset = Post.objects.filter(published_date__isnull=True).order_by('created_date')
+    context_object_name = 'posts'
+    template_name = 'blog/post_draft_list.html'
+
+
+class PublishPost(Protected, TemplateView):
```
Comment (webknjaz) [inline blog/views.py]: Inheritance order is incorrect.

File: blog/views.py [hunk 3]
```diff
@@ -38,7 +37,7 @@ class PostDetail(DetailView):
     template_name = 'blog/post_detail.html'
 
 
-class NewPost(CreateView, Protected):
+class NewPost(FormView, Protected):
```
Comment (webknjaz) [inline blog/views.py]: This breaks redirection.

File: blog/views.py [hunk 4]
```diff
@@ -2,13 +2,12 @@
 from django.utils import timezone
 from django.contrib.auth.decorators import login_required
 from django.utils.decorators import method_decorator
-from django.views.generic import (
-    DetailView, ListView,
-    UpdateView, View, CreateView,
-)
+from django.shortcuts import get_object_or_404, redirect
```
Comment (webknjaz) [inline blog/views.py]: You don't need these.

File: blog/views.py [hunk 5]
```diff
@@ -56,3 +55,70 @@ class EditPost(UpdateView, Protected):
...
+
+class RemovePost(Protected, DeleteView):
+    """View for deleting posts."""
+
+    model = Post
+    success_url = reverse_lazy('post_list')
+    template_name = 'blog/post_edit.html'
+
+
+class AddCommentToPost(FormView):
```
Comment (webknjaz) [inline blog/views.py]: Use `CreateView`

File: blog/views.py [hunk 6]
```diff
@@ -56,3 +55,70 @@ class EditPost(UpdateView, Protected):
...
+    """If someone wants to create new comment he/she get this view."""
+
+    form_class = CommentForm
+    template_name = 'blog/post_edit.html'
+
+    def post(self, request, *args, **kwargs):
```
Comment (webknjaz) [inline blog/views.py]: Drop this

File: blog/views.py [hunk 7]
```diff
@@ -56,3 +55,70 @@ class EditPost(UpdateView, Protected):
...
+            return redirect('post_detail', pk=post.pk)
+
+
+class ApproveComment(Protected, TemplateView):
+    """Moderate comment."""
+
+    def get(self, request, *args, **kwargs):
+        """Approve comment and update info in DB."""
+        comment = get_object_or_404(Comment, pk=kwargs['pk'])
+        comment.approve()
```
Comment (webknjaz) [inline blog/views.py]: According to `HTTP` spec, `GET` method is idempotent, i.e. it MUSTN'T change the resource it represents (or trigger any other actions within system). You should use either of `PUT`, `PATCH` or `POST` for this purpose.

File: blog/views.py [hunk 8]
```diff
@@ -56,3 +55,70 @@ class EditPost(UpdateView, Protected):
...
+        post = get_object_or_404(Post, pk=kwargs['pk'])
+        form = CommentForm(request.POST)
+        if form.is_valid():
+            comment = form.save(commit=False)
+            comment.post = post
+            comment.save()
+            return redirect('post_detail', pk=post.pk)
+
+
+class ApproveComment(Protected, TemplateView):
```
Comment (webknjaz) [inline blog/views.py]: TemplateView shouldn't be used for redirects, but for showing some representation.

File: blog/views.py [hunk 9]
```diff
@@ -56,3 +55,70 @@ class EditPost(UpdateView, Protected):
...
+        return redirect('post_detail', pk=comment.post.pk)
+
+
+class RemoveComment(Protected, DeleteView):
+    """Remove comment."""
+
+    model = Comment
+    template_name = 'blog/post_edit.html'
+
+    def post(self, request, *args, **kwargs):
```
Comment (webknjaz) [inline blog/views.py]: Drop this, let DeleteView handle logic.

File: blog/tests/test_models.py [hunk 10]
```diff
@@ -29,7 +29,31 @@ def test_post_publish_method(self):
         self.assertEqual(self.test_post.published_date, datetime(day=1, month=4, year=2016,
```
Comment (webknjaz) [inline blog/tests/test_models.py:29]: Where's teardown of this test suite?

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: blog/tests/test_models.py (hunk 10 above)
Comment (makkaronis4e) [reply blog/tests/test_models.py:29]: Done

**Review by webknjaz — APPROVED** [review]

**Pull request merged**